cpp/src/string/nvcategory_util.hpp

```diff
@@ -52,4 +52,13 @@ gdf_error sync_column_categories(gdf_column * input_columns[],gdf_column * outpu
  * @param[in,out] destination_table Contains columns that contain indices that map into source_table dictionaries.
  */
 gdf_error nvcategory_gather_table(cudf::table source_table, cudf::table destination_table);
+
+/**
+ * @brief Takes an input_column whose `values` of the category might be different from its input_column->data.
+ * Sync the `values` with the `data`, and remove unused keys in the category.
+ * The output_column should take ownership of the new category object created within this function.
+ * @param[in] input_column The input column
+ * @param[out] output_column The output column
+ */
+gdf_error clear_column_categories(gdf_column* input_column, gdf_column* output_column);
```

Comment:
`nvcategory_gather` directly replace the the `nvstring_category` of the `output_column` with that of the `input_column`. This works for gather: everything in the `output_column` replace something from the `input_column`. For scatter (although we call gather inside scatter) this is not the case. Some rows of the `output_column` are kept unchanged. Thus we need to first merge the two `nvstring_category`, update the `data` since the category has been changed in both columns, which is done by calling `sync_column_categories`, then perform the scatter. After the scatter though, some keys of the category may be used: so this `clear_column_categories` removes those unused keys and again update the `data`.